### PR TITLE
gettransaction: add an argument to decode the transaction

### LINKED ISSUE #16181: wallet: decode raw transaction in gettransaction
Does it make sense to decode the raw transaction in the wallet-rpc `gettransaction` call (or add an optional parameter to do so)?

See discussion in https://github.com/bitcoin/bitcoin/pull/15159#issuecomment-499051813

### PR DESCRIPTION
**Pull request by darosior**
This PR adds a new parameter to the `gettransaction` call : `decode`. If set to `true`, it will add a new `decoded` field to the response. This mimics the behavior of `getrawtransaction`'s `verbose` argument to avoid using 2 calls if we want to decode a wallet transaction (`gettransaction` then `decoderawtransaction`).

Fix #16181 .

### CODE REVIEW

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -1725,10 +1728,15 @@ static UniValue gettransaction(const JSONRPCRequest& request)
     uint256 hash(ParseHashV(request.params[0], "txid"));
 
     isminefilter filter = ISMINE_SPENDABLE;
-    if(!request.params[1].isNull())
+    if (!request.params[1].isNull())
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: Unrelated change, remove?

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -1725,10 +1728,15 @@ static UniValue gettransaction(const JSONRPCRequest& request)
...
         if(request.params[1].get_bool())
             filter = filter | ISMINE_WATCH_ONLY;
 
+    bool decode_tx = false;
+    if (!request.params[2].isNull())
+        if (request.params[2].get_bool())
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: `decode_tx = request.params[2].get_bool()`

**Review by promag — COMMENTED** [review]
Concept ACK, please include tests and a release note.

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -1664,13 +1664,14 @@ static UniValue gettransaction(const JSONRPCRequest& request)
...
 
-    if (request.fHelp || request.params.size() < 1 || request.params.size() > 2)
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 3)
         throw std::runtime_error(
             RPCHelpMan{"gettransaction",
                 "\nGet detailed information about in-wallet transaction <txid>\n",
                 {
                     {"txid", RPCArg::Type::STR, RPCArg::Optional::NO, "The transaction id"},
                     {"include_watchonly", RPCArg::Type::BOOL, /* default */ "false", "Whether to include watch-only addresses in balance calculation and details[]"},
+                    {"decode", RPCArg::Type::BOOL, /* default */ "true", "Whether to add a field with the decoded transaction"},
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: Default to false?

File: src/wallet/rpcwallet.cpp (hunk 1 above)
Comment (darosior) [reply src/wallet/rpcwallet.cpp]: Thought it would be great to fix this nit while passing through, will remove this change

File: src/wallet/rpcwallet.cpp (hunk 3 above)
Comment (darosior) [reply src/wallet/rpcwallet.cpp]: Yes, corrected.

**Comment by darosior** [comment]
@promag thank you for the review. I added the changes and will include the tests and release note soon.

**Comment by darosior** [comment]
> Concept ACK, please include tests and a release note.

Since there was no functional test for the `gettransaction` I created a new one. However it only tests the `decoded` field since I don't know if testing other would not be overkill.

File: test/functional/rpc_gettransaction.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env python3
```
Comment (promag) [inline test/functional/rpc_gettransaction.py]: Maybe add the test in test/functional/wallet_basic.p instead?

File: test/functional/rpc_gettransaction.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+    def run_test(self):
+        self.test_gettransaction_decoding()
+    
+    def test_gettransaction_decoding(self):
+        self.log.info("Testing gettransaction decoding...")
+        self.nodes[0].generate(101)
+        txid = self.nodes[0].sendtoaddress(self.nodes[0].getnewaddress(), 1)
+        # False for 'include_watchonly', True for 'decode'
+        tx = self.nodes[0].gettransaction(txid, False, True)
```
Comment (promag) [inline test/functional/rpc_gettransaction.py]: You can use named arguments and drop the comment above (untested):
```py
tx = self.nodes[0].gettransaction(txid=txid, decode=True)
```

Comment (darosior) [reply test/functional/rpc_gettransaction.py]: Yes, tested it it works thanks :D

File: test/functional/rpc_gettransaction.py (hunk 4 above)
Comment (darosior) [reply test/functional/rpc_gettransaction.py]: Ok.

**Comment by darosior** [comment]
Moved the test to [`test/functional/wallet_basic.py`](https://github.com/bitcoin/bitcoin/blob/master/test/functional/wallet_basic.py) and used named arguments.

File: test/functional/test_runner.py [hunk 6]
```diff
@@ -198,6 +198,7 @@
     'feature_blocksdir.py',
     'feature_config_args.py',
     'rpc_help.py',
+    'rpc_gettransaction.py',
```
Comment (fanquake) [inline test/functional/test_runner.py]: Please remove this entry now that the test is in `wallet_basic.py`.

**Comment by darosior** [comment]
Rebased.

**Comment by darosior** [comment]
Rebased.

File: src/wallet/rpcwallet.cpp [hunk 7]
```diff
@@ -1670,11 +1671,13 @@ static UniValue gettransaction(const JSONRPCRequest& request)
             "    ,...\n"
             "  ],\n"
             "  \"hex\" : \"data\"         (string) Raw data for transaction\n"
+            "  \"decoded\" : \"transaction\"         (json object) Optional, the decoded transaction\n"
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: Remove `\"transaction\"` as the value is not a string. Maybe mention that it's the same result as `decoderawtransaction`?

File: src/wallet/rpcwallet.cpp [hunk 8]
```diff
@@ -1693,6 +1696,10 @@ static UniValue gettransaction(const JSONRPCRequest& request)
         if(request.params[1].get_bool())
             filter = filter | ISMINE_WATCH_ONLY;
 
+    bool decode_tx = false;
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: nit, `const bool decode = request.params[2].isNull() ? false : request.params[2].get_bool();`

File: src/wallet/rpcwallet.cpp [hunk 9]
```diff
@@ -1693,6 +1696,10 @@ static UniValue gettransaction(const JSONRPCRequest& request)
...
+    if (!request.params[2].isNull())
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: nit, otherwise add `{ }`.

File: test/functional/wallet_basic.py [hunk 10]
```diff
@@ -496,6 +496,13 @@ def run_test(self):
         self.nodes[0].setlabel(change, 'foobar')
         assert_equal(self.nodes[0].getaddressinfo(change)['ischange'], False)
 
+        # Test "decoded" field value in gettransaction response
+        self.log.info("Testing gettransaction decoding...")
```
Comment (promag) [inline test/functional/wallet_basic.py]: nit, remove log.

File: test/functional/wallet_basic.py [hunk 11]
```diff
@@ -496,6 +496,13 @@ def run_test(self):
...
+        self.nodes[0].generate(1)
+        txid = self.nodes[0].sendtoaddress(self.nodes[0].getnewaddress(), 1)
```
Comment (promag) [inline test/functional/wallet_basic.py]: nit, remove and use the last transaction below?

File: test/functional/wallet_basic.py [hunk 12]
```diff
@@ -496,6 +496,13 @@ def run_test(self):
...
+        tx = self.nodes[0].gettransaction(txid=txid, decode=True)
+        assert_equal(tx["decoded"], self.nodes[0].decoderawtransaction(tx["hex"]))
```
Comment (promag) [inline test/functional/wallet_basic.py]: Or you can simply change L488:
```diff
-tx = self.nodes[0].decoderawtransaction(self.nodes[0].gettransaction(txid)['hex'])
+tx = self.nodes[0].gettransaction(txid, decode=True)['decoded']
```

File: test/functional/wallet_basic.py (hunk 10 above)
Comment (darosior) [reply test/functional/wallet_basic.py]: Is there a rationale for removing the log ? I added it to mimic the other tests' behavior

File: src/wallet/rpcwallet.cpp (hunk 7 above)
Comment (darosior) [reply src/wallet/rpcwallet.cpp]: done

File: src/wallet/rpcwallet.cpp (hunk 8 above)
Comment (darosior) [reply src/wallet/rpcwallet.cpp]: done

File: test/functional/wallet_basic.py (hunk 11 above)
Comment (darosior) [reply test/functional/wallet_basic.py]: done

File: test/functional/wallet_basic.py (hunk 12 above)
Comment (darosior) [reply test/functional/wallet_basic.py]: Probably better to leave it as a separated test

**Comment by darosior** [comment]
Some minor corrections according to last review.

**Comment by darosior** [comment]
Some minor corrections according to last review.

**Review by promag — COMMENTED** [review]
You could use https://github.com/bitcoin/bitcoin/commit/0b6ba66238c377116bc6c21e19cffbf1b6dfc788#diff-0f596a69ccfb4d07c007cd6fc993898dR50 (not yet in master) to do:
```py
assert_no_key('decoded', self.nodes[0].gettransaction(txid, decode=False))
```

File: test/functional/wallet_basic.py [hunk 13]
```diff
@@ -496,6 +496,11 @@ def run_test(self):
         self.nodes[0].setlabel(change, 'foobar')
         assert_equal(self.nodes[0].getaddressinfo(change)['ischange'], False)
 
+        # Test "decoded" field value in gettransaction response
```
Comment (jb55) [inline test/functional/wallet_basic.py]: this is a weird nit, but just curious why there's a "Non-breaking space" character between the `#` and `Test` (utf-8 0xC2A0)

```
git cat-file -p d5ab374c48a7f2708acc5e122044a3a9cfc1ac6a | grep '#.Test "decoded"' | xxd | head -n1
```

File: doc/release-notes.md [hunk 14]
```diff
@@ -99,6 +99,10 @@ Low-level Changes section below.
   `-limitancestorcount`, `-limitdescendantcount` and `-walletrejectlongchains`
   command line arguments.
 
+* The `gettransaction` RPC now accepts a third (boolean) argument `decode`. If
+  set to `true`, a new `decoded` field will be added to the response containing
```
Comment (jb55) [inline doc/release-notes.md]: There's another one of those chars here after `decoded` and before `field`. Not a big deal, just was wondering what editor you are using that would be adding these randomly?

```
git cat-file -p a2b23686c65d7e8c7e69b7547322d2eea1cf3b44 | xxd | grep c2a0
00000f90: 6465 6460 c2a0 6669 656c 6420 7769 6c6c  ded`..field will
```

**Review by jb55 — COMMENTED** [review]
Concept ACK

File: doc/release-notes.md (hunk 14 above)
Comment (darosior) [reply doc/release-notes.md]: Vim, you right it's weird

File: test/functional/wallet_basic.py (hunk 13 above)
Comment (darosior) [reply test/functional/wallet_basic.py]: I think it's caused by a vim module (deoplete-jedi on this one I think)

**Comment by darosior** [comment]
Rebased.
*(And addressed the spacing nits pointed by @jb55 :-) )*

File: doc/release-notes-16185.md [hunk 15]
```diff
@@ -0,0 +1,3 @@
+RPC changes
+-----------
+The `gettransaction` RPC now accepts a third (boolean) argument `decode`. If set to `true`, a new `decoded` field will be added to the response containing
```
Comment (achow101) [inline doc/release-notes-16185.md]: I think the sentence got cut off here.

File: src/wallet/rpcwallet.cpp [hunk 16]
```diff
@@ -1724,6 +1729,12 @@ static UniValue gettransaction(const JSONRPCRequest& request)
     std::string strHex = EncodeHexTx(*wtx.tx, pwallet->chain().rpcSerializationFlags());
     entry.pushKV("hex", strHex);
 
+    if (decode_tx) {
+        UniValue decoded(UniValue::VOBJ);
+        TxToUniv(*wtx.tx, uint256(), decoded, false, pwallet->chain().rpcSerializationFlags());
```
Comment (achow101) [inline src/wallet/rpcwallet.cpp]: nit: It's not necessary to have `pwallet->chain().rpcSerializationFlags()`.

**Review by achow101 — COMMENTED** [review]
Concept ACK

**Comment by darosior** [comment]
@achow101 thank you for the review, force pushed to rectify.

**Comment by achow101** [comment]
ACK a636daba4f61c32ec83d6677b8cff66dbde3fda0 Reviewed code and tested locally

File: src/wallet/rpcwallet.cpp [hunk 17]
```diff
@@ -1676,11 +1677,13 @@ static UniValue gettransaction(const JSONRPCRequest& request)
             "    ,...\n"
             "  ],\n"
             "  \"hex\" : \"data\"         (string) Raw data for transaction\n"
+            "  \"decoded\" : transaction         (json object) Optional, the decoded transaction\n"
```
Comment (luke-jr) [inline src/wallet/rpcwallet.cpp]: This seems wanting... but expanding here would be ugly.

**Review by luke-jr — APPROVED** [review]
utACK

**Review by meshcollider — COMMENTED** [review]
utACK a636daba4f61c32ec83d6677b8cff66dbde3fda0

Once rebased, I will merge this

**Comment by darosior** [comment]
Rebased.

**Comment by promag** [comment]
You must make linters happy.

**Comment by darosior** [comment]
> You must make linters happy.

No that's a Travis error : `curl: (7) Failed to connect to storage.googleapis.com port 443: Connection timed out`

**Review by meshcollider — COMMENTED** [review]
re-utACK 9965940e35c445ccded55510348af228ff22f0e9

File: src/wallet/rpcwallet.cpp [hunk 18]
```diff
@@ -1684,11 +1685,13 @@ static UniValue gettransaction(const JSONRPCRequest& request)
...
```
Comment (MarcoFalke) [inline src/wallet/rpcwallet.cpp]: Should mention that the field is identical to `decoderawtransaction`?

**Comment by jnewbery** [comment]
Sorry for the post-merge review comment, but would it make sense to call the new parameter `verbose`, to mirror the `verbose` argument in `getrawtransaction`?

If so, we should make that change before v0.19 to avoid an API change.

**Comment by ryanofsky** [comment]
> Sorry for the post-merge review comment, but would it make sense to call the new parameter `verbose`, to mirror the `verbose` argument in `getrawtransaction`?
> 
> If so, we should make that change before v0.19 to avoid an API change.

This seems like a good idea. You could open a one-line PR and add the milestone to get more feedback.

**Comment by promag** [comment]
+1 for API consistency.
